plasmapy/diagnostics/charged_particle_radiography.py

```diff
@@ -1115,118 +1263,149 @@ def max_deflection(self):
-                    np.max(np.abs(self.grid.pts0.to(u.m).value)),
-                    np.max(np.abs(self.grid.pts1.to(u.m).value)),
-                    np.max(np.abs(self.grid.pts2.to(u.m).value)),
-                ]
-            )
+    optical_density: `bool`
+        If `True`, return the optical density rather than the intensity
+
+        .. math::
+            OD = -log_{10}(Intensity/I_0)
+
+        where :math:`Intensity` is the simulation intensity on the
+        detector plane and :math:`I_0` is the intensity on the detector
+        plane in the absence of simulated fields. Default is `False`.
+
+    Returns
+    -------
+    hax : `~astropy.units.Quantity` array shape ``(hbins,)``
+        The horizontal axis of the synthetic radiograph in meters.
+
+    vax : `~astropy.units.Quantity` array shape ``(vbins, )``
+        The vertical axis of the synthetic radiograph in meters.
+
+    intensity : `~numpy.ndarray`, shape ``(hbins, vbins)``
+        The number of particles counted in each bin of the histogram.
+    """
 
-            # The factor of 5 here is somewhat arbitrary: we just want a
-            # region a few times bigger than the image of the grid on the
-            # detector, since particles could be deflected out
-            size = 5 * np.array([[-w, w], [-w, w]]) * u.m
+    # results_dict will raise an error if the simulation has not been run.
+    if isinstance(obj, Tracker):
+        d = obj.results_dict
 
-        # Generate the histogram
-        intensity, h, v = np.histogram2d(
-            xloc, yloc, range=size.to(u.m).value, bins=bins
+    # Check if dictioanry-like (dict or npz file)
+    elif isinstance(obj, dict):
+        d = obj
+    else:
+        raise ValueError(
+            "The first argument of synthetic_radiograph must be "
+            "either a cpr.Tracker or an "
+            "output dictionary from cpr.Tracker"
         )
 
-        # h, v are the bin edges: compute the centers to produce arrays
-        # of the right length (then trim off the extra point)
-        h = ((h + np.roll(h, -1)) / 2)[0:-1]
-        v = ((v + np.roll(v, -1)) / 2)[0:-1]
+    if bins is None:
+        bins = [200, 200]
 
-        # Throw a warning if < 50% of the particles are included on the
-        # histogram
-        percentage = np.sum(intensity) / self.nparticles
-        if percentage < 0.5:
-            warnings.warn(
-                f"Only {percentage:.2%} of the particles are shown "
-                "on this synthetic radiograph. Consider increasing "
-                "the size to include more.",
-                RuntimeWarning,
-            )
+    # Note that, at the end of the simulation, all particles were moved
+    # into the image plane.
+
+    # If ignore_grid is True, use the predicted positions in the absence of
+    # simulated fields
+    if ignore_grid:
+        xloc = d["x0"]
+        yloc = d["y0"]
+    else:
+        xloc = d["x"]
+        yloc = d["y"]
+
+    if size is None:
```

Comment:
Should add some `size` validations here to handle types and shape.  Something like...

```python
if size is None:
    ...
elif not isinstance(size, np.ndarray):
    raise TypeError
elif size.shape != (2, 2):
    raise ValueError
elif not isinstance(size, u.Quantity):
    warn("Assuming SI")
    size = size * u.m
```

Reply:
Note that `Quantity` is a sub-class of `ndarray` so `isinstance(5 * u.m, np.ndarray)` will be `True`.

Reply:
Done, except I just kept the requirement that `size` be an explicit `u.Quantity`. Since detectors vary in size (but are always much smaller than a meter), allowing raw numbers interpreted as SI units is an unlikely use case.